armory/instrument/instrument.py

```diff
@@ -0,0 +1,670 @@
+        if keep != "all":
+            raise NotImplementedError(f"keep={keep}, use 'all'")
+        self.keep = keep
+        self.records = []
+        # TODO: checking
+
+    def set_filepath(self, filepath):
+        self.filepath = filepath
+
+    def _write(self, name, batch, result):
+        self.records.append((name, batch, result))
+
+    def collate_results(self):
+        """
+        Return a map from name to output, in original order.
+        """
+        output = {}
+        for name, batch, result in self.records:
+            if name not in output:
+                output[name] = []
+            output[name].append(result)
+        return output
+
+    def close(self):
+        output = self.collate_results()
+        with open(self.filepath, "w") as f:
+            json.dump(output, f)
+        # TODO: push to results dictionary?
+
+
+# GLOBAL CONTEXT METHODS #
+
+
+def get_context():
+    """
+    Get the context state object for the experimental procedure
+    """
+    global _CONTEXT
+    if _CONTEXT is None:
+        _CONTEXT = Context()
+    return _CONTEXT
+
+
+def get_probe(name: str = ""):
+    """
+    Get a probe with specified name, creating it if needed
+    """
+    if name != "" and not str.isidentifier(name):
+        raise ValueError(f"name {name} should be an identifier or the empty string")
+
+    if name not in _PROBES:
+        probe = Probe(name)
+        probe.set_sink(get_context())
+        _PROBES[name] = probe
+    return _PROBES[name]
+
+
+def connect_meter(meter):
+    get_context().connect_meter(meter)
+
+
+def add_meter(*args, **kwargs):
+    meter = Meter(*args, **kwargs)
+    connect_meter(meter)
+
+
+def get_meters():
+    return get_context().get_meters()
+
+
+def add_writer(writer):
+    get_context().add_writer(writer)
+
+
+def main():
+    """
+    Just for current WIP demonstration
+    """
+    # Begin model file
+    import numpy as np
+
+    # from armory.instrument import get_probe
+    probe = get_probe("model")
+
+    class Model:
+        def __init__(self, input_dim=100, classes=10):
+            self.input_dim = input_dim
+            self.classes = classes
+            self.preprocessor = np.random.random(self.input_dim)
+            self.predictor = np.random.random((self.input_dim, self.classes))
+
+        def predict(self, x):
+            x_prep = self.preprocessor * x
+            # if pytorch Tensor: probe.update(lambda x: x.detach().cpu().numpy(), prep_output=x_prep)
+            probe.update(lambda x: np.expand_dims(x, 0), prep_output=x_prep)
+            logits = np.dot(self.predictor.transpose(), x_prep)
+            return logits
+
+    # End model file
+
+    # Begin metric setup (could happen anywhere)
+
+    # from armory.instrument import add_writer, add_meter
+    from armory.utils import metrics
+
+    add_writer(PrintWriter())
+    add_meter(
+        "postprocessed_l2_distance",
+        metrics.l2,
+        "model.prep_output[benign]",
```

Comment:
similar to [this](https://github.com/twosixlabs/armory/pull/1387/files#r851465631) comment, the brackets-in-a-string notation I found a bit non-intuitive. It's not immediately clear to me what the possible values that could go inside the brackets are. The different stages?

Relatedly, the `process_meter_arg()` function feels potentially brittle. If I call

```
 add_meter(
        "postprocessed_l2_distance",
        metrics.l2,
        "model.prep_output[[benign]",
        "model.prep_output[adversarial]",
    )
```

I get a not-so-clear error:

```
 File "/workspace/armory/instrument/instrument.py", line 617, in main
    "model.prep_output[adversarial]",
  File "/workspace/armory/instrument/instrument.py", line 570, in add_meter
    connect_meter(meter)
  File "/workspace/armory/instrument/instrument.py", line 565, in connect_meter
    get_context().connect_meter(meter)
  File "/workspace/armory/instrument/instrument.py", line 267, in connect_meter
    self.mapper.connect_meter(meter)
  File "/workspace/armory/instrument/instrument.py", line 187, in connect_meter
    probe_variable, stage_filter = process_meter_arg(arg)
  File "/workspace/armory/instrument/instrument.py", line 148, in process_meter_arg
    arg, filt = arg.split("[")
ValueError: too many values to unpack (expected 2)
```

Reply:
The latter was an actual error in my input checking ("and" instead of "or"), so that now the error is:
```
>>> instrument.process_meter_arg("model.prep_output[[benign]")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/david.slater/git/davidslater/armory/armory/instrument/instrument.py", line 160, in process_meter_arg
    raise ValueError(f"arg '{arg}' must have a single matching '[]' pair or none")
ValueError: arg 'model.prep_output[[benign]' must have a single matching '[]' pair or none
```

Reply:
Currently, it is just different stages that go inside the brackets. There is a larger question of how to enable more expressive filters.